### ISSUE TITLE
'UndocumentedApi: public members of private nested classes are incorrectly treated as public API'

### ISSUE THREAD

**Issue by RED-FLAG-DEVELOPMENT** [SonarOpenCommunity/sonar-cxx#3107]
---
name: Bug report
about: Create a report to help us improve
title: 'UndocumentedApi: public members of private nested classes are incorrectly treated as public API'
labels: ''
assignees: ''

---

**Before you create a new issue**
- This appears specific to the cxx plugin's `cxx:UndocumentedApi` rule rather than a general SonarQube issue.
- The sonar-cxx Wiki was reviewed.
- The issue was reproduced with debug information enabled using `-X -Dsonar.verbose=true`.
- Existing sonar-cxx issues were searched; no duplicate was found.

**Describe the bug**

`cxx:UndocumentedApi` reports a public member of a private nested class as public API. In the following code, `Foo::Impl` is private, so `Foo::Impl::bar()` is not externally accessible API:

```cpp
// Minimal sonar-cxx cxx:UndocumentedApi reproduction.
class Foo
{
private:
    class Impl
    {
    public:
        void bar();
    };
};
```

The analysis reports exactly one issue in this file:

- `include/repro/nested_private.hpp:8`: `Undocumented API: bar`

The private nested class `Impl` itself is not reported.

**To Reproduce**

1. Check out the public reproduction repository at commit [`b4cff0786a128082f543c5a6ec2a211987c9a794`](https://github.com/CB-st/sonar-cxx-undocumented-api-repro/tree/b4cff0786a128082f543c5a6ec2a211987c9a794).
2. Use a C++ quality profile with `cxx:UndocumentedApi` active.
3. Run SonarScanner with debug logging enabled: `-X -Dsonar.verbose=true`.
4. Inspect the issues for `include/repro/nested_private.hpp` and `include/repro/nested_public.hpp`.
5. Observe the issue on the private nested class member at `nested_private.hpp:8`.

The public-nested control behaves as expected and reports:

- `include/repro/nested_public.hpp:5`: `Undocumented API: Impl`
- `include/repro/nested_public.hpp:8`: `Undocumented API: bar`

**Expected behavior**

`Foo::Impl::bar()` should not be reported by `cxx:UndocumentedApi` because its enclosing class, `Foo::Impl`, is private. The public control should continue to report both the public nested class and its public member.

**Screenshots**

Not applicable; the exact issue locations and messages are listed above.

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04, Linux x64
 - SonarQube version: 26.7.0.124771
 - cxx plugin version: 2.2.2, build 1409
 - sonar-scanner version: 8.0.1.6346

**Additional context**

A likely cause may be in `AbstractCxxPublicApiVisitor.isPublicApiMember`: it appears that public visibility of the member may be considered without also requiring all enclosing nested classes to be publicly accessible. This is only a hypothesis based on the observed behavior.

**Comment by guwirth** [comment]
Hello @RED-FLAG-DEVELOPMENT,

Thank you for your feedback!

> A likely cause may be in AbstractCxxPublicApiVisitor.isPublicApiMember: it appears that public visibility of the member may be considered without also requiring all enclosing nested classes to be publicly accessible. This is only a hypothesis based on the observed behavior.

https://github.com/SonarOpenCommunity/sonar-cxx/blob/012faa50ac87f8ad6eeb31c277e136637bf9cec5/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java#L245

Usage is then like this:

```c++
class Foo
{
private:
    class Impl
    {
    public:
        void bar();
    };

    // Foo can own an objekt (or a pointer) from Impl
    Impl myImpl; 

public:
    void useBar() 
    {
        // Foo calls the method of the inner class
        myImpl.bar(); 
    }
};
```
Why do you think description of Impl is not necessary?

Regards,

**Comment by guwirth** [comment]
Whether class Impl should be documented depends entirely on who reads the documentation and where the class is defined.

Here is the golden rule for C++:

## 1. No, in the Public API Documentation
Since Impl is in the private section of Foo, it is a strict implementation detail. End-users (other developers who just want to use Foo) cannot access or instantiate it.

* Clutter reduction: Keeping it out of public docs keeps the API clean and easy to understand.
* Automation: Documentation tools like Doxygen disable the generation of private members by default for this exact reason.

## 2. Yes, as Internal Code Comments
Developers who maintain or extend the code of Foo need to understand what Impl does. You should document it directly in the source code (or header file) with a focus on why it exists.

* What to document:
  * Responsibility: What part of Foo does it handle? (e.g., "Encapsulates platform-specific logic for Windows/Linux").
  * Rationale: Why was it separated? (e.g., "Prevents heavy third-party headers from leaking into Foo.h").

------------------------------
## Best Practice Example
If you use an automated tool like Doxygen, you can use the @internal tag. This keeps the comments in the code but hides them from the public customer-facing documentation:

```c++
class Foo
{private:
    /**
     * @internal
     * @class Impl
     * @brief Encapsulates the internal data processing for Foo.
     * 
     * This class is private to keep the public ABI of Foo stable
     * and isolate internal dependencies from the header file.
     */
    class Impl
    {
    public:
        /** @brief Performs the actual heavy calculation for Foo::execute(). */
        void bar();
    };
};
```

Summary: Document Impl directly in the code for yourself and your team, but keep it hidden from the public API documentation.

**Comment by ConnorS-P** [comment]
@guwirth you Raise a valid point about internal documentation, but you could use the same logic to argue ALL code should receive documentation. The issue here is that a private method and a public method on a private internal class sit on the same level of internal privacy, one trips the rule and the other does not.

In theory a dev working in that space would need knowledge of both these things.

I think the question is less of "should the code be documented" and more of "does this rule as written enforce an atomic and orthogonal metric of code quality". By your own example I feel it does not; a team could care primarily about internal documentation, external documentation, or both. I feel the rule should either support some configuration to allow the distinction, or be split into 2 rules.

Currently I feel all consumers are under served.

Documentation absolutists do not get private methode coverage.

Those only concerned with public facing docs wind up getting flagged for internal documentation gaps.

You also run into the truly bizzare requirement where the methods of the class must be documented, but the class itself need not.

**Comment by guwirth** [comment]
@ConnorS-P I'm sure you're right, both ways have advantages and disadvantages. My main concerns are that this rule has existed for a long time and the previous users seem to be satisfied with it. If we change that, the previous users are unhappy.

> I feel the rule should either support some configuration to allow the distinction, or be split into 2 rules.

I'm not a big fan of configurations. I also wonder how an additional rule should work exactly. How many other cases are there with nested access specifiers that we also have to consider?

**Comment by ConnorS-P** [comment]
"Satisfied with" and "tolerant of" can be difficult to differentiate in the absence of feedback. For what it's worth, I am a user and I am unsatisfied. 🙂

Do you have data on how many people just mark the results of this as a false positive?

**Comment by ConnorS-P** [comment]
> I'm not a big fan of configurations. I also wonder how an additional rule should work exactly.

In my mind, there are two reasonably distinct consumers with two straightforward quality goals.

Say we have Alice and Bob.

Bob maintains an open-source project with contributors and external reviewers. He wants someone without deep domain knowledge to be able to understand and reason about the implementation.

Bob wants a rule whose sentiment is roughly:

> Implementation should be documented.

For Bob, documentation coverage should include public, protected, and private implementation details. If a private class or private member function is important enough to exist, it is potentially important enough for a maintainer to understand.

Alice maintains a closed-source library used internally by other teams. Her company's policy requires teams to publish documentation for the libraries they expose to other teams, and her team generates that documentation automatically with Doxygen.

Alice wants a rule whose sentiment is roughly:

> API exposed to consumers should be documented.

For Alice, the relevant scope is the externally meaningful API: public interfaces and, where applicable, protected extension points. Private implementation details are deliberately outside that contract.

The current rule does not fully satisfy either user.

Alice is unhappy because her team spends time writing API-style documentation for entities her consumers cannot access and which will never appear in their public-facing documentation or getting bogged down with marking multiple false positives.

Bob is unhappy because the rule does not actually enforce documentation of the implementation. A contributor can satisfy it while leaving private member functions and private internal classes undocumented. In the nested-class case, the current behavior is especially unusual: the private class itself does not require documentation, its private members do not require documentation, but its public members do.

That seems internally inconsistent regardless of which documentation philosophy you prefer.

I can see two relatively simple ways to model the distinction.

**Option 1: Two rules**

One rule could enforce documentation of externally relevant API using normal C++ accessibility semantics.

A second rule could enforce documentation of implementation details, including private declarations.

Either rule could optionally allow a regex or similar pattern to define the expected documentation structure/style if teams want to enforce more than simply "a Doxygen comment exists."

**Option 2: One rule with an accessibility threshold**

The existing rule could expose a simple setting for the minimum accessibility level requiring documentation:

* Public
* Protected
* Private

Separately, the rule could allow teams to define the expected documentation format or structure, for example via a regex.

That does not seem like configuration for configuration's sake to me. The accessibility setting determines "which declarations are in scope", while the documentation-pattern setting determines "what acceptable documentation looks like".

My main concern with the current behavior is that it mixes the two models. It is neither "document the externally relevant API" nor "document the implementation." Instead, whether an internal declaration requires documentation can depend on the access specifier immediately surrounding that declaration, even when its enclosing type makes it inaccessible to the consumer.

That is the inconsistency I would like to see resolved, however you choose to go about it.

> How many other cases are there with nested access specifiers that we also have to consider?

On this I have no idea. I am by no means a C++ language expert and potentially many more corner cases could exist.

**Comment by guwirth** [comment]
I will test the following sample again. For the existing rule both should at least behave in the same way.

```c++
class Foo {
private:
    class Impl1 {
    public:
        void bar();
    };
    struct Impl2 {
        void bar();
    };
};
```

**Comment by guwirth** [comment]
Here are the test files from the original author of the rule:
https://github.com/SonarOpenCommunity/sonar-cxx/tree/master/cxx-checks/src/test/resources/checks/UndocumentedApiCheck

**Comment by ConnorS-P** [comment]
@guwirth Just following up on this, what did you learn from the retest/test files?

**Comment by guwirth** [comment]
@ConnorS-P sorry had no time to lock into it. I’m on vacation…
